%\documentclass{report}
%\documentclass[11pt,twoside,a4paper]{report}
\documentclass[11pt,journal,compsoc]{report}[1]
\usepackage[letterpaper, margin=0.75in]{geometry}
\usepackage{lipsum}
\usepackage{epstopdf}
\usepackage{amsthm}
\PassOptionsToPackage{hyphens}{url}
\usepackage[T1]{fontenc}
\usepackage[draft]{hyperref}
\usepackage{graphicx}
\usepackage[cmex10]{amsmath}
%\usepackage{algorithmic}
%\usepackage{algorithm}
\usepackage{array}
\usepackage{mdwmath}
\usepackage{mdwtab}
\usepackage{eqparbox}
\usepackage{ucs}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
%\usepackage{graphicx}
%\usepackage{caption}
\usepackage{url}
\usepackage{multirow}
\usepackage{hhline}
\usepackage{multicol}
\usepackage{xcolor}
\usepackage{listings}
\lstset{basicstyle=\ttfamily,
  showstringspaces=false,
  commentstyle=\color{red},
  keywordstyle=\color{blue}
}

%\definecolor{coback}{RGB}{245,245,245}
%
%\lstdefinestyle{myCustomMatlabStyle}{
%  language=c,
%  numbersep=10pt,
%  tabsize=4,
%  showspaces=false,
%  showstringspaces=false,
%  backgroundcolor=\color{coback},
%  keywordstyle=\color{black}\bfseries
%} 

\usepackage[framemethod=tikz]{mdframed}

\definecolor{light-gray}{gray}{0.95}
\lstset{basicstyle=\scriptsize\upshape\ttfamily,tabsize=4,numbersep=10pt,language=Matlab,keywordstyle=\color{black}\bfseries}

\surroundwithmdframed[
  hidealllines=true,
  backgroundcolor=light-gray,
  innerleftmargin=10pt,
  innertopmargin=0pt,
  innerbottommargin=0pt]{lstlisting}                  

\DeclareMathOperator*{\argmin}{argmin}
\DeclareMathOperator*{\argmax}{argmax}
\newcommand*{\argminl}{\argmin\limits}
\newcommand*{\argmaxl}{\argmax\limits}
\newcommand{\Am}{\mathrm{A}}
\newcommand{\Cm}{\mathrm{C}}
\newcommand{\Gm}{\mathrm{G}}
\newcommand{\Tm}{\mathrm{T}}


\usepackage[pagestyles]{titlesec}
\titleformat{\chapter}[display]{\normalfont\bfseries}{}{0pt}{\huge}
\newpagestyle{mystyle}
{\sethead[\thepage][][\chaptertitle]{}{}{\thepage}}
\pagestyle{mystyle}

%\copyrightyear{2016} \pubyear{2016}
%\access{Advance Access Publication Date: Day Month 2016}
%\appnotes{Manuscript Category}

\AtBeginDocument{\renewcommand{\abstractname}{}}
\renewcommand{\abstractname}{} 
\renewcommand{\baselinestretch}{1.3}

\begin{document}
%\firstpage{1}

\title{
%
\begin{figure}[h!]
\centerline{\includegraphics[width=5cm]{../imgs/logo.pdf}}
\label{logo}
\end{figure}
~\\
\textbf{A toolkit for DNA sequence\\ analysis and manipulation}
~\\~\\
\large
D. Pratas (pratas@ua.pt)\\
A. J. Pinho (ap@ua.pt)
~\\~\\
\small
IEETA/DETI, University of Aveiro, Portugal\\
~\\
Version 1.7.17
}
\date{}
\maketitle

\tableofcontents

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\chapter*{1. Introduction}
\addcontentsline{toc}{chapter}{1. Introduction}
\label{intro}

Recent advances in {DNA} sequencing have revolutionized the field of genomics,
making it possible for research groups to generate large amounts of sequenced
data, very rapidly and at substantially lower cost. Its storage have been
made using specific file formats, such as FASTQ and FASTA. Therefore, its
analysis and manipulation is crucial \cite{Buermans-2014a}. Several
frameworks for analysis and manipulation emerged, namely GALAXY
\cite{Giardine-2005a}, GATK \cite{DePristo-2011a}, HTSeq
\cite{Anders-2014a}, MEGA \cite{Kumar-2016a}, among others.
In the majority, these frameworks require licenses and do not provide
a low level access to the information, since they are commonly approached
by scripting or interfaces.

We present GOOSE, a (free) novel toolkit for analyzing and manipulating
SEQ (``ACGTN''), FASTA or FASTQ formats, with many complementary tools,
for Linux-based systems, built for fast processing.
GOOSE supports pipes for easy integration. It includes tools for information 
display, randomizing, edition, conversion, extraction, searching,
calculation and visualization. GOOSE is prepared to deal high
very large datasets, typically GB and TB sizes. For this dimension of data
fast access and flexibility is crucial, which are some of the qualities of
GOOSE. 

The toolkit is a command line version, using the prefix ``goose-'' 
followed by the suffix with the respective name of the program.
GOOSE is implemented in C language and it is freely available, under GPLv3, 
at \url{https://pratas.github.io/goose}.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\chapter*{2. FASTQ tools}
\addcontentsline{toc}{chapter}{2. FASTQ tools}
\label{fastq}

Current available tools for FASTQ format analysis and manipulation include:
\begin{enumerate}

\item \textbf{goose-fastq2fasta}
\item \textbf{goose-fastq2mfasta}
\item \textbf{goose-fastqclustreads}
\item \textbf{goose-FastqExcludeN}
\item \textbf{goose-FastqExtractQualityScores}
\item \textbf{goose-FastqInfo}
\item \textbf{goose-FastqMaximumReadSize}
\item \textbf{goose-FastqMinimumLocalQualityScoreForward}
\item \textbf{goose-FastqMinimumLocalQualityScoreReverse}
\item \textbf{goose-FastqMinimumQualityScore}
\item \textbf{goose-FastqMinimumReadSize}






\item \textbf{goose-count}
\item \textbf{goose-extract}
\item \textbf{goose-extractreadbypattern}



\item \textbf{goose-fastqpack}
\item \textbf{goose-fastqsimulation}
\item \textbf{goose-FastqSplit}
\item \textbf{goose-FastqTrimm}
\item \textbf{goose-fastqunpack}
\item \textbf{goose-filter}
\item \textbf{goose-findnpos}

\item \textbf{goose-genrandomdna}
\item \textbf{goose-getunique}
\item \textbf{goose-info}
\item \textbf{goose-mfmotifcoords}

\item \textbf{goose-mutatedna}
\item \textbf{goose-mutatefasta}
\item \textbf{goose-mutatefastq}
\item \textbf{goose-newlineonnewx}
\item \textbf{goose-period}
\item \textbf{goose-permuteseqbyblocks}

\item \textbf{goose-randfastaextrachars}
\item \textbf{goose-randfastqextrachars}
\item \textbf{goose-randseqextrachars}
\item \textbf{goose-real2binthreshold}
\item \textbf{goose-reducematrixbythreshold}
\item \textbf{goose-renamehumanheaders}
\item \textbf{goose-reverse}
\item \textbf{goose-reverselm}
\item \textbf{goose-searchphash}
\item \textbf{goose-segment}
\item \textbf{goose-seq2fasta}
\item \textbf{goose-seq2fastq}
\item \textbf{goose-SequenceToGroupSequence}
\item \textbf{goose-splitreads}

\item \textbf{goose-wsearch}
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\chapter*{3. FASTA tools}
\addcontentsline{toc}{chapter}{3. FASTA tools}
\label{fasta}

\begin{enumerate}
\item \textbf{goose-fasta2seq}
\item \textbf{goose-fastaextract}
\item \textbf{goose-fastainfo}
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\chapter*{4. Sequence tools}
\addcontentsline{toc}{chapter}{4. Sequence tools}
\label{seq}

Current available sequence tools for analysis and manipulation include:
\begin{enumerate}
\item \textbf{goose-AminoAcidToGroup}
\item \textbf{goose-ProteinToPseudoDNA}
\item \textbf{goose-geco}
\item \textbf{goose-gede}
\end{enumerate}

\section*{4.1 Program goose-AminoAcidToGroup}
\addcontentsline{toc}{section}{4.1~~~Program goose-AminoAcidToGroup}

The goose-AminoAcidToGroup converts an amino acid sequence to a group sequence.
In the following subsections, we explain the input and output paramters.

\subsection*{Input parameters}
\addcontentsline{toc}{subsection}{4.1.1~~~Input parameters}

The goose-AminoAcidToGroup program needs two streams for the computation,
namely the input and output standard. The input stream is an amino acid sequence.
The attribution is given according to:
\begin{lstlisting}
Usage: ./goose-AminoAcidToGroup < in.prot > out.group
It converts a amino acid sequence to a group sequence.
Table:
Prot	Group
R	P
H	P  Amino acids with electric charged side chains: POSITIVE
K	P
-	-
D	N
E	N  Amino acids with electric charged side chains: NEGATIVE
-	-
S	U
T	U
N	U  Amino acids with electric UNCHARGED side chains
Q	U
-	-
C	S
U	S
G	S  Special cases
P	S
-	-
A	H
V	H
I	H
L	H
M	H  Amino acids with hydrophobic side chains
F	H
Y	H
W	H
-	-
*	*  Others
X	X  Unknown
\end{lstlisting}
It can be used to group amino acids by properties, such as electric charge (positive
and negative), uncharged side chains, hydrophobic side chains and special cases.

\subsection*{Output}
\addcontentsline{toc}{subsection}{4.1.2~~~Output}

The output of the goose-AminoAcidToGroup program is a group sequence.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\chapter*{5. General purpose tools}
\addcontentsline{toc}{chapter}{5. General purpose tools}
\label{seq}

\begin{enumerate}
\item \textbf{goose-comparativemap}: visualisation of comparative maps. It builds a image given specific patterns between two sequences.
\item \textbf{goose-BruteForceString}: it generates, line by line, multiple combinations of strings up to a certain size.
\item \textbf{goose-char2line}: it transforms each char into a char in each line.
\item \textbf{goose-sum}: it adds the second column value to the first column value.
\item \textbf{goose-min}: it finds the minimum value between two column values.
\item \textbf{goose-minus}: it substracts the second column value to the first column value.
\item \textbf{goose-max}: it finds the mmaximum value between two column values.
\end{enumerate}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\chapter*{Note 11: Software availability \& characteristics}
\addcontentsline{toc}{chapter}{Note 11: Software availability \& characteristics}
\label{C-SOFTWARE}

~\\
~\\
All source code for the FALCON is available at:
\begin{itemize}
\begin{sloppypar}
\item \url{https://github.com/pratas/falcon}.
\end{sloppypar}
\end{itemize}
~\\
FALCON includes the following programs (See Note~2 for a common pipeline using 
the programs):
\begin{enumerate}
\begin{sloppypar}
\item \textbf{FALCON} - infers metagenomic composition;
\item \textbf{FALCON-FILTER} - filters local similarity of inferred sequences;
\item \textbf{FALCON-EYE} - displays the output from FALCON and FALCON-FILTER in 
a compact map;
\end{sloppypar}
\end{enumerate}

\section*{FALCON program}
\addcontentsline{toc}{section}{11.1~~~FALCON program}

The FALCON program enables to identify and quantify the similarity between
present-days reference sequences and the FASTQ reads from ancient genomes.
In the following subsections, we explain the input and output paramters.

\subsection*{Input parameters}
\addcontentsline{toc}{subsection}{11.1.1~~~Input parameters}

The FALCON program needs two files for the computation. The FILE1,
with the reads provided from the NGS platform, and the FILE2,
with the multi-FASTA cointaining the sequences of the genomes and
respective headers, are mandatory arguments. The rest of the parameters
are non mandatory, namely:
\begin{lstlisting}
Usage: FALCON [OPTION]... [FILE1] [FILE2]                                
A compression-based method to infer metagenomic sample composition.      
                                                                         
Non-mandatory arguments:                                                 
                                                                         
  -h                       give this help,                               
  -F                       force mode (overwrites top file),             
  -V                       display version number,                       
  -v                       verbose mode (more information),              
  -Z                       database local similarity,                    
  -s                       show compression levels,                      
  -l <level>               compression level [1;44],                    
  -p <sample>              subsampling (default: 1),                    
  -t <top>                 top of similarity (default: 20),              
  -n <nThreads>            number of threads (default: 2),              
  -x <FILE>                similarity top filename,                      
  -y <FILE>                local similarities filename,                  
                                                                         
Mandatory arguments:                                                     
                                                                         
  [FILE1]                  metagenomic filename (FASTA or FASTQ),        
  [FILE2]                  database filename (FASTA or Multi-FASTA).
\end{lstlisting}
There are two hidden parameters. One for setting the cache size:
\begin{lstlisting}
  -c <cache>             maximum collisions for hash cache. Memory     
                         values are higly dependent of the parameter   
                         specification.                                
\end{lstlisting}
This is a parameter that affects the precision and computational resources
needed for the computation. Lower values use less memory. For a complete
description see \cite{Pratas-2016a,Pratas-2016b}.
The previous parameter is only needed when the context of the Markov model
is higher than 13. For setting specifically the models through the command
line use the following explanation (more information can be found at 
\cite{Pratas-2016a,Pratas-2016b}):

\subsection*{Output}
\addcontentsline{toc}{subsection}{11.1.2~~~Output}

The output of the FALCON program are two files. One file contains the top-$n$ 
values with the highest similarity relatively to the reads. The following 
output shows an example of a top-5:
\begin{lstlisting}
1	9472	92.597	548558394	Human endogenous retrovirus K113
2	4668621	42.184	1069460419	Escherichia coli strain 210221272
3	4558287	42.016	1008930592	Shigella sp. PAMC 28760
4	4574246	36.661	844762407	Shigella boydii strain ATCC 9210
5	4878853	36.531	992379426	Shigella sonnei strain FDAARGOS 90
\end{lstlisting} 
These column values stand, respectively, for the ranking of higher similarity, 
the sequence size, the percentage of similarity, the global identifyer (GI) and
the name of the FASTA read (genome name).\\
~\\
The other file contains the relative local similarities and it is only 
available when FALCON runs with the ``-Z'' flag. This file is the input of
the FALCON-FILTER program that we describe next. This file is packed in 
a compact format for not increasing the substantially the storage.

\section*{FALCON-FILTER program}
\addcontentsline{toc}{section}{11.2~~~FALCON-FILTER program}

The FALCON-FILTER program enables to identify and quantify where the 
similarity between present-days reference sequences and the FASTQ reads 
from ancient genomes, below a certain threshold, occurs.
In the following subsections, we explain the input and output paramters.

\subsection*{Input parameters}
\addcontentsline{toc}{subsection}{11.2.1~~~Input parameters}

The FALCON-FILTER program needs only one file for the computation. The
file is provided from the output of the FALCON program (when the flag
``-Z'' is set on the FALCON program). The non-mandatory arguments are
mostly filters and the threshold.
\begin{lstlisting}
Usage: FALCON-FILTER [OPTION]... [FILE]                                  
Filter and segment FALCON output.                                        
                                                                         
Non-mandatory arguments:                                                 
                                                                         
  -h                       give this help,                               
  -F                       force mode (overwrites top file),             
  -V                       display version number,                       
  -v                       verbose mode (more information),              
  -s  <size>               filter window size,                           
  -w  <type>               filter window type,                           
  -x  <sampling>           filter window sampling,                       
  -sl <lower>              similarity lower bound,                       
  -su <upper>              similarity upper bound,                       
  -dl <lower>              size lower bound,                             
  -du <upper>              size upper bound,                             
  -t  <threshold>          threshold [0;2.0],                            
  -o  <FILE>               output filename,                              
                                                                         
Mandatory arguments:                                                     
                                                                         
  [FILE]                   profile filename (from FALCON).
\end{lstlisting}

\subsection*{Output}
\addcontentsline{toc}{subsection}{11.2.2~~~Output}

The output of the file contains the coordinates of the relative
similar regions with the respective self-similarity classification.

\addcontentsline{toc}{chapter}{Bibliography}
\bibliographystyle{IEEEtran}
\bibliography{defs,SPLc,SPLj,Cod,Bio,Book,Anc,Patho}

\end{document}
